### DISCUSSION TITLE
New Scenario Analysis

### DISCUSSION THREAD

**Discussion by jafluri** [ZEN-universe/ZEN-garden#294] [Functionalities]
# Overview

This post introduces the new scenario analysis in ZEN-Garden along with all its features and some examples. The main topics are:

- How to define scenarios in the new format
- Hierarchical expansion of sets to quickly define parameter changes for multiple elements
- How to run scenarios in parallel on Euler
- Defining parameter values in lists to avoid lengthy manual definitions

## Defining a scenario

In the new format, scenarios are defined in the following way:

```python
scenarios = {"scenario_name_1": 
                {"element_1": 
                    {"param_1":
                        {"default": "<attribute_file_name>",
                         "default_op": <float>,  
                         "file": "<file_name>",
                         "file_op": <float>,
                        }
                    }
                }
            }
```

As before, each scenario has a name and the parameters from the different elements can be overwritten to perform a different analysis. However, the new format highlights reusability. For each parameter, one can set the following four keys:
- "default": Change the file from which the default value is read out (with or without ending ".csv")
- "default_op": A factor that will be applied to the default value
- "file": Change the file name from which the remaining values are read (with or without ending ".csv")
- "file_op": A factor that is applied to the final data frame, including the default values

For example, to overwrite the "heat" element one could do something like

```python
scenarios["5"] = {"heat": {"demand": {"file": "demand_5",
                                      "file_op": 1.5,
                                      "default": "attributes_v2",
                                      "default_op": 0.25}}}
```

Note that not all four keys have to be present, and any subset is valid. Missing keys are automatically replaced with the default values. Further, one can use all files and values in multiple scenario analyses, e.g. the file "attributes_v2" could be used in scenario "4" as well. 

**Important:** The "file_op" is applied after the data frame has been created, so it will be applied to the default values as well. This means that setting both "default_op" and "file_op" will change the default values twice.

### Overwriting entire sets

In some cases, one would like to change a parameter for all elements in a set. To do this one can use the same syntax for set names, e.g.

```python
scenarios["6"] = {"set_technologies": {"max_load": {"file": "max_load_5",
                                                  "file_op": 1.5,
                                                  "default": "attributes_v2",
                                                  "default_op": 0.25,
                                                  "exclude": ["tech1", "tech2"]}}}
```

For sets, an additional key "exclude" is allowed, which allows one to define a list of elements in the list that should not be overwritten. The set expansion works hierarchical, meaning that if one defines the same parameter manually for an element of the set, this parameter is not touched **at all**. So even if some keys are not manually defined, but in the set definition, they won't be set. For example, let's say we have `set_technologies = ["tech1", "tech2"]` and 
```python
scenarios["7"] = {"set_technologies": {"max_load": {"file": "max_load_5"}},
                             "tech1": {"max_load": {"default": 3}}}
```
the final scenario dictionary after expansion would be
```
{"tech1": {"max_load": {"default": 3}},
 "tech2": {"max_load": {"file": "max_load_5"}}}
```
This hierarchy is continued for smaller sets, e.g. defining "set_transport_technologies" takes precedence to "set_technologies", etc.

### Defining parameters with lists

Additionally, it is possible to define parameters in lists, to conduct a sensitivity analysis. For example

```python
scenarios["5"] = {"heat": {"demand": {"file": "demand_5",
                                      "file_op": 1.5,
                                      "default": "attributes_v2",
                                      "default_op": [0.25, 0.3, 0.35]}}}
```

Will create 3 new scenarios for all values in the "default_op". All keys support this, however, it is **important** that the value is a proper Python list, not an array or something else, so please wrap the values in `list(...)` if you generate the iterable with `np.linspace0`,  `range` or similar. If multiple lists are defined in the same scenario, all possible combinations (cartesian product) are generated, so watch out for combinatorial explosions. 

The default names for the generated scenarios are `"p{i:02d}_{j:03d}"`, where `i` is an int referring to the expanded parameter name (e.g. heat, demand, file, default_op) and `j` to its value in the list (e.g. [0.25, 0.3, 0.35]). The mappings of `i` and `j` to the parameter names and values are written to a "param_map.yml" in the root directory of the corresponding scenario (see below). It is possible to overwrite this default naming with a formatting key:

```python
scenarios["5"] = {"heat": {"demand": {"file": "demand_5",
                                      "file_op": 1.5,
                                      "default": "attributes_v2",
                                      "default_op": [0.25, 0.3, 0.35],
                                      "default_op_fmt": "heat_demand_{}"}}}
```

The formatting key is the original key containing the list followed by "_fmt". The value of the formatting key has to be a string containing the format literal "{}". The formatting string "{}" will then be replaced by each of the values of the list. For example here, we would generate the three scenarios "heat_demand_0.25", "heat_demand_0.3" and "heat_demand_0.35".

Defining parameters in lists changes the output format of ZEN-Garden. Previously, all scenarios were in a single output folder. Now when a scenario contains one or multiple lists, all sub-scenarios are also in a subfolder, for example, the output structure could look something like this:

```
dataset_1/
    scenario_1/
    scenario_2/
        scenario_p00_000_p001_000/
        scenario_p01_000_p001_000/
        ...
        param_map.yml
    scenario_3/
...
```

Here, "scenario_2" was defined via lists and its sub-scenarios are now in subfolders with the definitions of the parameters in the "param_map.yml". Folders containing sub-scenarios can be read in with the `Results` class of ZEN-Garden. For example,

```python
from zen_garden.postprocess.results import Results

all_scenarios = Results("dataset_1")
sub_scenarios = Results("dataset_1/scenario_2")
```

Here `all_scenarios` will read in all scenarios of the run and sub-scenarios can be accessed by appending them to the name, e.g. `all_scenarios.get_df("capacity", scenario="scenario_2_p00_000_p001_000")`. While `sub_scenarios` will read only the sub-scenarios of "scenario_2" and the different sub-scenarios can be accessed directly, e.g. `sub_scenarios.get_df("capacity", scenario="scenario_p00_000_p001_000")` (note the missing "2" for the base-scenario name).

### Using both, sets and lists

When using both, set and list expansion, list expansion is done first. For example

```python
scenarios["6"] = {"set_technologies": {"max_load": {"file_op": [1.5, 2.5, 3.5],
                                                  "exclude": ["tech1", "tech2"]}}}
```

will only generate 3 scenarios where the "file_op" for all technologies (except ["tech1", "tech2"]) are set to the values in the lists simultaneously.

### Overwriting Analysis and System

It is also possible to overwrite entries in the system and analysis dictionaries. The syntax is straightforward:

```python
scenarios["5"] = {"system": {key: value},
                  "heat": {"demand": {"file": "demand_5",
                                      "file_op": 1.5,}}}
```

Note that there is a strict type check when overwriting the system or analysis, i.e. the value used for "value" **must** have the same type as the value already in the dictionary.

## Running the analysis

Per default, all scenarios are run sequentially, as before. Additionally, one can specify a subset of scenarios to run with the `--job_index` argument. For example

```bash
python -m zen_garden --job_index 1,4,7
```

will run scenarios 1,4,7, where the number is the index of the key (starting with 0), not the key itself (no explicit scenario names).

Finally, when submitting a job on the Euler-cluster with --array=start-stop:step%Nmax (start and stop are inclusive, Nmax is the max number of concurrent jobs) or other --array options (e.g. --array=1,4,7), each job will run only one index, automatically. Note that the indices start again with 0, so running the first four scenarios would be --array=0-3 (per default the step is 1 and Nmax default to the number of submitted jobs). If you do not submit the job with the --array argument, all scenarios are run sequentially as in the local case.

**Comment by jacob-mannhardt** [comment]
If you want to change the yearly variation of a time-dependent input data, i.e., adding a file for `demand_yearly_variation`, you can proceed as follows:
`scenarios["high"] = {"electricity": {"demand_yearly_variation":{"file":"demand_yearly_variation_high"}}}`

Note that you overwrite the `demand_yearly_variation` parameter, not `demand`.
